igvc_navigation/src/path_follower/Smooth_control.cpp

```diff
@@ -0,0 +1,243 @@
+#include "Smooth_control.h"
+
+void Smooth_control::get_trajectory(igvc_msgs::velocity_pair& vel, nav_msgs::PathConstPtr path,
+                                    nav_msgs::Path& trajectory, RobotState state, RobotState& target)
+{
+  RobotState start_state = state;
+
+  unsigned int path_index;
+  Eigen::Vector3d los;
+  Eigen::Vector3d heading;
+
+  path_index = get_closest_position(path, state);
+  target.setState(get_target_position(path, path_index, state));
+  get_los_and_heading(los, heading, state, target);
+  Egocentric_angles angles = get_egocentric_angles(path, path_index, los, heading);
+
+  // distance to target
+  double tar_dist = state.distTo(target);
+  // calculate timestep
+  double est_time = tar_dist / m_target_velocity;
+  double dt = est_time / m_granularity;
+
+  for (int i = 0; i < m_granularity; i++)
```

Comment:
It just hit me that if `granularity` is <1, then no velocity commands are generated, meaning that visualization would interfere with functionality. Consider adding the line:

```C++
m_granularity = std::max(1, m_granularity);
```